analysis/complex.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2017 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kevin Buzzard
+
+The complex numbers, modelled as R^2 in the obvious way.
+
+TODO: Add topology, and prove that the complexes are a topological ring.
+-/
+import analysis.real
+noncomputable theory
+
+local attribute [instance] classical.decidable_inhabited classical.prop_decidable
+
+structure complex : Type :=
+(re : ℝ) (im : ℝ)
+
+notation `ℂ` := complex
+
+namespace complex
+
+theorem eta (z : complex) : complex.mk z.re z.im = z :=
+  cases_on z (λ _ _, rfl)
+
+theorem eq_of_re_eq_and_im_eq (z w : complex) : z.re=w.re ∧ z.im=w.im → z=w :=
+begin
+intro H,rw [←eta z,←eta w,H.left,H.right],
+end
+
+-- simp version
+
+theorem eq_iff_re_eq_and_im_eq (z w : complex) : z=w ↔ z.re=w.re ∧ z.im=w.im :=
+begin
+split,
+  intro H,rw [H],split;trivial,
+exact eq_of_re_eq_and_im_eq _ _,
+end
+
+lemma proj_re (r0 i0 : real) : (complex.mk r0 i0).re = r0 := rfl
+lemma proj_im (r0 i0 : real) : (complex.mk r0 i0).im = i0 := rfl
+
+local attribute [simp] eq_iff_re_eq_and_im_eq proj_re proj_im
+
+def of_real : ℝ → ℂ := λ x, { re := x, im := 0 }
```

Comment:
This should be `def of_real (r : ℝ) : ℂ := { re := x, im := 0 }`
This is not only an aesthetic reason (why use the unnecessary lambda?), but a practical one:
With `of_real : ℝ → ℂ` the generated simp rule for `of_real` matches all occurrences of `of_real`, while the later only matches `of_real r`.

Reply:
This I really do not understand -- I mean, I understand that you're asking me to change something, and I understand how to change it, but I do not understand why changing something to something which looks to me to be definitionally equal should be a sensible idea. Of course I do not doubt that it is! But I would really like to understand better where these comments are coming from. What is "the generated simp rule for of_real"?

Reply:
You should try `#print prefix of_real` with both versions. The generated `*.equations` lemma for `of_real` will have an `x` in Johannes's version on the LHS, meaning that if you `simp [of_real]` it will not unfold unless it's already been applied to an argument.

Reply:
My slow understanding of these comments is currently at the stage where I understand that even though the two approaches give definitionally equal functions, the associated equation lemmas are not the same, however I am unclear about why one prefers Johannes' version, because I am not clear about what is unfolding or whether we want it to unfold. This, as is sometimes the case with me, seems to come back to the fact that I do not actually understand what simp is doing or how it does it. I guess that's my point here -- the two equation lemmas are "different ways of saying exactly the same thing" to me, but for someone who knows better what simp is doing I guess this is not the case.

Reply:
Let me try the following explanation:
First, in this case **unfold** just means, that `simp` will rewrite occurrences `of_real` using the equation provided by the definition.

Second, it is hard to say why I prefer the `of_real r` variant. At first glance the `of_real : R -> C` variant would match more often. But by experience I had quiet some proofs where I **only** want to replace `f x` instead of every occurrence of `f`.